### PR TITLE
Add timeline read markers API

### LINKED ISSUE #4093: Endpoint for last read notification
When using multiple clients (such as web and mobile), it would be nice to have a count of unread notifications that is synced between both.

There are many ways this could be implemented. A simple solution off the top of my head is an endpoint for "ID of last read notification" which you can POST and maybe GET, and and "unread": field on Notification objects that is set to `id > lastReadId` (that check could be done locally, of course, but that would require 2 GETs instead of one for each update).

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

### LINKED ISSUE #7418: Endpoint for last read toot
(Somewhat similar idea to #4093)

It would be nice to have a synced timeline across clients (web or not) connected to an instance. This functionality suggestion is directly inspired by the (discontinued) [Tweet Marker](https://github.com/manton/tweetmarker) 3rd party service for Twitter.

It would allow clients to set & get the last viewed toot on the instance itself, thus allowing them to know where to continue from when displaying the timeline. When using multiple clients it would make switching between them easier since each of the clients could show the same point in the instance's timeline.

I cannot give implementation suggestions but maybe the [Tweet Marker API description](https://github.com/manton/tweetmarker/blob/master/documentation/v2.md) could be a helpful starting point.

* * * *
- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by Gargron**
Fix #4093, fix #7418, close #6756

|Method|Description|
|-----|----|
|`GET /api/v1/markers`|Get markers specified by `timeline` array param. Response is a map with timeline names as keys|
|`POST /api/v1/markers`|Create new marker(s) with `last_read_id` by submitting a map with timeline names as keys. The keys can be `home` or `notifications`|

JSON returned includes `last_read_id`, `updated_at` and `version` (which increments on every update).

### CODE REVIEW

File: app/javascript/mastodon/actions/markers.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    params.notifications = {
+      last_read_id: lastNotificationId,
+    };
+  }
+
+  // At least in Firefox, the request does not send unless
+  // console.log is used like this
+
+  const beacon = sendBeacon(accessToken, params);
+  console.log('Saving timeline markers...', beacon);
```
Comment (ykzts) [inline app/javascript/mastodon/actions/markers.js]:
```suggestion
  console.log('Saving timeline markers...', beacon); // eslint-disable-line no-console
```

File: app/javascript/mastodon/actions/markers.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import api from '../api';
```
Comment (ykzts) [inline app/javascript/mastodon/actions/markers.js]: unused?

**Review by ykzts — CHANGES_REQUESTED** [review]

**Review by ykzts — APPROVED** [review]

**Comment by nolanlawson** [comment]
I have a few questions about this implementation:

1. It seems the Mastodon webapp will submit the markers in the unload event. So if a user keeps their tab open in their browser and never closes it, another client will never know what the last read position is? It seems like it may be better to do this lazily (e.g. every few minutes), as well as during the unload event and when a page goes into a background tab.
2. This is only for the home/notifications timelines right now, if I understand correctly?
3. Is the assumption that, on first load, a client will immediately scroll to the saved position in the timeline if a marker is found? So if a user has not checked Mastodon in several days and there are 100s of statuses, should the app load all 100+ of them and then scroll to the position? Or give up and reset the position? Or lazily show the new content?

/cc @charlag

**Comment by charlag** [comment]
I agree with @nolanlawson on the first one.
I think when you open it on another client you just load:
 1. The latest ones
 2. Statuses with max_id = marker and 
and put a loadable gap between them (I think we do something like that already).
cc @connyduck

**Comment by Gargron** [comment]
> It seems like it may be better to do this lazily (e.g. every few minutes), as well as during the unload event and when a page goes into a background tab.

Well, what if you have multiple clients open? And they all try to update the position at the same time? If they all use the streaming API I suppose they would arrive at similar or identical results but it seems wasteful to just keep updating it back and forth between them all. I'm not sure what the point would be of saving and loading a marker while a client is already actively open, the other app would then load the top of the timeline anyway..?

**Comment by charlag** [comment]
It could be done in the way that we send it to the server and server writes it only when thr client stops communicating but that's complicated I guess.
I dot think anyone will load marker when the app is already open, only on startup. If two clients open, the last used one is likely to win (which makes sense to me).

**Pull request merged**